viamonvif/discovery.go

```diff
@@ -0,0 +1,354 @@
+		} `xml:"GetStreamUriResponse"`
+	} `xml:"Body"`
+}
+
+// getDeviceInformationResponse is the schema the GetDeviceInformation response is formatted in.
+type getDeviceInformationResponse struct {
+	XMLName xml.Name `xml:"Envelope"`
+	Body    struct {
+		GetDeviceInformationResponse struct {
+			Manufacturer string `xml:"Manufacturer"`
+			Model        string `xml:"Model"`
+			SerialNumber string `xml:"SerialNumber"`
+		} `xml:"GetDeviceInformationResponse"`
+	} `xml:"Body"`
+}
+
+// CameraInfo holds both the RTSP URLs and supplementary camera details.
+type CameraInfo struct {
+	RTSPURLs     []string `json:"rtsp_urls"`
+	Manufacturer string   `json:"manufacturer"`
+	Model        string   `json:"model"`
+	SerialNumber string   `json:"serial_number"`
+}
+
+// CameraInfoList is a struct containing a list of CameraInfo structs.
+type CameraInfoList struct {
+	Cameras []CameraInfo `json:"cameras"`
+}
+
+// DiscoverCameras performs WS-Discovery using the use-go/onvif discovery utility,
+// then uses ONVIF queries to get available RTSP addresses and supplementary info.
+func DiscoverCameras(username, password string, logger logging.Logger) (*CameraInfoList, error) {
+	var discoveredCameras []CameraInfo
+	ctx, cancelFunc := context.WithCancel(context.Background())
+	defer cancelFunc()
+
+	interfaces, err := net.Interfaces()
+	if err != nil {
+		return nil, fmt.Errorf("failed to retrieve network interfaces: %w", err)
+	}
+
+	// Filter out unsuitable interfaces
+	var validInterfaces []net.Interface
+	for _, iface := range interfaces {
+		if iface.Flags&net.FlagUp != 0 && iface.Flags&net.FlagMulticast != 0 && iface.Flags&net.FlagLoopback == 0 {
+			validInterfaces = append(validInterfaces, iface)
+		} else {
+			logger.Debugf("skipping interface %s: does not meet WS-Discovery requirements", iface.Name)
+		}
+	}
+	// If no valid interfaces, return an error
+	if len(validInterfaces) == 0 {
+		return nil, errors.New("no valid net interfaces for WS-Discovery")
+	}
+
+	resultsChan := make(chan []CameraInfo, len(validInterfaces))
+	activeWorkers := sync.WaitGroup{}
+	for _, iface := range validInterfaces {
+		activeWorkers.Add(1)
+		utils.ManagedGo(func() {
+			logger.Debugf("sending WS-Discovery probe using interface: %s", iface.Name)
+			var discoveryResps []string
+			iterations := 3 // run ws-discovery probe 3 times due to sync flakiness between announcer and requester
+			for i := range make([]int, iterations) {
+				if ctx.Err() != nil {
+					logger.Infof("context canceled for interface: %s", iface.Name)
+					return
+				}
+
+				// other ws-discovery args
+				scopes := []string{}
+				types := []string{"dn:NetworkVideoTransmitter"}
+				namespaces := map[string]string{}
+				resp, err := wsdiscovery.SendProbe(iface.Name, scopes, types, namespaces)
+				if err != nil {
+					logger.Warnf("attempt %d: failed to send WS-Discovery probe on interface %s: %v", i+1, iface.Name, err)
+					break
+				}
+				discoveryResps = append(discoveryResps, resp...)
```

Comment:
[q] There are no issues with having duplicate responses here from multiple probes?

Reply:
No duplicates. Checked manually and with AI